src/app/qgshandlebadlayers.cpp

```diff
@@ -552,6 +559,11 @@ void QgsHandleBadLayers::autoFind()
     const QString longName = dataInfo.fileName();
     QString provider = node.namedItem( QStringLiteral( "provider" ) ).toElement().text();
     const QString fileType = mLayerList->item( i, 2 )->text();
+
+    progressDialog.setValue( i );
+    progressDialog.setLabelText( QObject::tr( "Searching for file %1. [ %2 of %n ]", nullptr, layersToFind.size() ).arg( name, i ) );
+    QCoreApplication::processEvents();
```

Comment:
`processEvents()` is dangerous and (unfortunately) can not be used without risking random crashes.

I would therefore leave this either as a pure feedback (I hope the bar is moving even without this call?) without the possibility to cancel.
In a future step you can also move things to a background task, and allow cancelling. But be prepared that this will be a thorough refactoring.

Reply:
Out of curiosity can you explain why processEvents is dangerous.

I did some test right now and everything seemed fine other than the cancel button was sometimes unresponsive to clicks but not to enter.

I was hoping that if this was not usable as is, though it is clearly suggested in the qt docs, we would have something in house that could work as an alternative, something like what is used in the attribute table (I forgot to check the code for that).

Reply:
[Here](https://github.com/qgis/QGIS/blob/209e9144f87654ac6dc11dadfc539d387d571ece/src/app/3d/qgs3danimationwidget.cpp#L209) the processEvents is used and the dialog cancel is linked to the feedback class. Is that usage better? Otherwise I could include a feedback class to pass in the finder function, this would also lead to a more prompt stop if the cancel button works properly.

Reply:
By doing a github search I can see that processEvents is used throughout QGIS, is it being phased out or kept in due to lack of alternatives?

Reply:
> Out of curiosity can you explain why processEvents is dangerous.

Very briefly: it processes events in unpredictable contexts (see e.g. this comment for a workaround https://github.com/qgis/QGIS/blob/master/src/gui/qgsmapcanvas.cpp#L566 , this discussion here ... https://github.com/qgis/QGIS-Enhancement-Proposals/issues/122#issuecomment-382887214)

> Here the processEvents is used and the dialog cancel is linked to the feedback class. Is that usage better?

No

> Otherwise I could include a feedback class to pass in the finder function, this would also lead to a more prompt stop if the cancel button works properly.

This will only work if threaded. And making this code threaded is involved and will likely prolong the review process and merge time.

> By doing a github search I can see that processEvents is used throughout QGIS, is it being phased out or kept in due to lack of alternatives?

Phased out slowly because writing safe, threaded code takest time

Reply:
I did some tests, without the process events, using either a progress bar in the base dialog or by simply removing it in this dialog.

I either case the UI just freezes and the progress bar doesn't even appear. In the case of the dialog only the window appear.

Other than multithread I don't think there is a way other than process events to keep the UI responsive.